### ISSUE TITLE
Separate testing for dataset fetchers

### ISSUE THREAD

**Issue by kchawla-pi** [nilearn/nilearn#2283]
The OSF.io servers which host our didactic datasets were erratic during December. It inconvenienced PRs, as tests would fail not because of the changes made to the code, which is the only reason a test should fail, but because the datasets couldn't be loaded.

We have tolerated this because we want the dataset fetchers & servers to be regularly tested so we are notified in case of problems. However, this has also caused test failures and frustrations. In may cases we don't have the ability to resolve the server failure, especially since moving all the dtaa to OSF which we don't have any influence over.

This is even more galling on CircleCI, since the doc builds take a long time.

It occurs to me that we don't have to subject ourselves and our contributors to the whims of this setup.

What I propose is we run tests for the fetchers separately. For testing the rest of the code and building documentation, we use cached versions of the datasets, instead of downloading them afresh.

This might be trickier to do with Travis and Azure Pipelines, I am not yet certain of their caching mechanisms, but it can be very easily implemented in CircleCI. At the very least, it'll ameliorate some of the frustrations. We can modify the other CTs as their caching methods evolve.

**Comment by kchawla-pi** [comment]
Seems Travis had and now Azure Pipelines has implemented better caching mechanisms than when last I checked. Good.

**Comment by emdupre** [comment]
Was wrestling with this in #2279. A series of empty commits worked for me, in case anyone else hits the same problem and wants a clean-ish approach before this lands !

**Comment by GaelVaroquaux** [comment]
In the unit tests (not the docs building) we should mock the downloads,
and the unit tests should run without access to internet. This was the
case for all our unit tests at some point. Is it not the case anymore?

That said, I see this as an argument to limit the number of datasets
that we support, and the number of different sources
(related: https://github.com/nilearn/nilearn/issues/2020 ).

**Comment by kchawla-pi** [comment]
Even so, servers fail, and sometimes it is days before they are fixed. They shouldn't interfere with a PR's development. I think my idea is sound. You decide and tell me.

> Is it not the case anymore?

Nope. I don't yet understand how the request mockers work for this. The current ones are due for a rewrite.  Too many globals.

> limit the number of datasets

Sure.
It may also mean changing many examples.
Will the didactic power of the examples might reduce, since we will no longer represent diverse datasets for students to cut their teeth on?

**Comment by GaelVaroquaux** [comment]
> I think my idea is sound. You decide and tell me.

The right solution is mocking. That's what it is there for.

>     Is it not the case anymore?

> Nope. I don't yet understand how the request mockers work for this. The current
> ones are due for a rewrite. Too many globals.

We need to fix them.

Maybe some of them are still working (those that where contributed a
while ago). You could try running the tests without internet to see which
one work and which one do not.

Once we have fixed those that do not, we can consider adding a CI
configuration that runs the tests without internet, so that the problem
does not reappear.

>     limit the number of datasets

> Sure. It may also mean changing many examples. Will the didactic power
> of the examples might reduce, since we will no longer represent diverse
> datasets for students to cut their teeth on?

Well, it's a tradeoff, as always. But we can kill a few (the nyu, the
adhd). It also means less code, which is good.

**Comment by kchawla-pi** [comment]
Sounds good to me. Mocking for the tests is good and we should fix it. I believe it is the recent tests that don't mock properly, the older ones are fine.

What about for docs building? Like I mentioned, the OSF servers were erratic for several weeks. This would insulate PRs from such events.

**Comment by GaelVaroquaux** [comment]
> What about for docs building? Like I mentioned, the OSF servers were erratic
> for several weeks. This would insulate PRs from such events.

We cannot build the example without the data.

What we could do would be to try to have good caching of this data on
circleCI. This would be beneficial.

**Comment by kchawla-pi** [comment]
That's what I was proposing.
I'll workout a detailed plan and update it here.

**Comment by GaelVaroquaux** [comment]
> That's what I was proposing.

Cool!

**Comment by jeromedockes** [comment]
> I'll workout a detailed plan and update it here.

Any news on this @kchawla-pi ?

**Comment by kchawla-pi** [comment]
I'll try to do this on the weekend.

**Comment by jeromedockes** [comment]
great, let us know if you need help!

**Comment by jeromedockes** [comment]
trying to revive this: general cleanup of datasets package

Nilearn now has a lot of fetchers and the datasets package is the part of the
library that fails the most, the platforms hosting the datasets are not 100%
reliable.

we could:

- track down any datasets that we can live without. If we could reduce a bit the
  diversity of datasets used in the example gallery and deprecate some fetchers
  that would be great. Let's make a list here.

- replace custom code using `urllib` with `requests`, which would be a new
  dependency. IMHO it is easier to use and more reliable. For example yesterday
  `fetch_localizer_button_task` was failing as downloading
  <https://osf.io/hwbm2/download> resulted in a truncated json file, but I could
  download the whole file with `requests`.

- as discussed above, systematically use mocking when testing all fetchers.
  Maybe we could also have a way of testing them without mocking and run that
  periodically -- but this is more or less what the `full-build` of the examples
  does on circleci.

- not sure if this makes sense, but if we identify some crucial datasets that
  often fail to be downloaded, maybe we could consider storing them on two
  platforms to have some redundancy.

**Comment by jeromedockes** [comment]
Also, we could rewrite and reduce the functionality of the neurovault fetcher.
(It doesn't cause any tests to fail, and it could cause an example to fail when
neurosynth was down but that has been fixed today.) Still, I think it is mostly
useful to download individual images and whole collections that have been
selected by the user, ie selected by ID. However it has been created with
meta-analysis in mind, filtering the downloaded images based on terms and values
in their metadata. If we simplify it by only allowing selecting images or whole
collections by ID we can remove a lot of code and address some issues such as
the completion of partially downloaded collections.

**Comment by jeromedockes** [comment]
One of the first things we could remove is downloading "decoding" terms from
neurosynth for neurovault images. It is used in only one example and even there
is not essential, and this API often fails. Such code should probably live in
the neurosynth package or a separate client rather than in nilearn.

**Comment by emdupre** [comment]
In terms of priorities, what do you think of first replacing urllib with requests code ? I'm wondering if it will stabilize the problematic fetchers in the short term and give us time to comb through the tests and mock more extensively.

**Comment by jeromedockes** [comment]
that makes sense, and probably would come with changes to the mocking utilities -- we will need to mock requests. 
using `requests` needs approval from more people eg @GaelVaroquaux and @bthirion though, as it adds a new dependency -- even it is a very lightweight, stable, and widely used one.

**Comment by kchawla-pi** [comment]
I believe Gael already green lit using requests.

**Comment by bthirion** [comment]
> One of the first things we could remove is downloading "decoding" terms from
> neurosynth for neurovault images. It is used in only one example and even there
> is not essential, and this API often fails. Such code should probably live in
> the neurosynth package or a separate client rather than in nilearn.

Ye, I think that this is somewhat unessential. the natural way to access Neurosynth is by dataset ID I think. I like this simplification, but we have to go through deprecation cycles.

**Comment by bthirion** [comment]
>     * track down any datasets that we can live without. If we could reduce a bit the
>       diversity of datasets used in the example gallery and deprecate some fetchers
>       that would be great. Let's make a list here.

* Cobre fetcher is never used if I'm not mistaken
* I'd like to remove th FIAC dataset download and use: it's not a great dataset.
* I'd be happy to get rid of mixed gambles.

**Comment by jeromedockes** [comment]
so for deprecation we have:
- [ ] cobre
- [ ] FIAC
- [ ] mixed gambles
- [ ] nyu
- [ ] adhd


I have two questions:
- do we do each in a separate PR? it might be easier as some will require rewriting examples
- what release number do we announce will remove them?

**Comment by bthirion** [comment]
> do we do each in a separate PR? it might be easier as some will require rewriting examples
Sounds like a good idea

> what release number do we announce will remove them?
0.9 I think

**Comment by jeromedockes** [comment]
@thomasbazeille which dataset could replace mixed gambles in the fREM example?

**Comment by thomasbazeille** [comment]
> @thomasbazeille which dataset could replace mixed gambles in the fREM example?

Actually there is not really, unless I'm not aware about some hidden decoding dataset. 

As much as I do love Haxby, I think we need something else to showcase Decoding / fREM and potentially also another example of decoding after GLM fitting for an event-related paradigm (#1506 is maybe related, and others as well)

Mixed Gambles is really impractical and does not give strong result so I won't mind removing it but I think we should have something better instead if we want to showcase nilearn full scope.

**Comment by jeromedockes** [comment]
@AlexandreAbraham do you know why `_fetch_file` can do authentication? I don't see this used anywhere

**Comment by jeromedockes** [comment]
it seems to have been added in this commit 
`29c9cc93b34fc3c9d94ff60bba18b379bf935e15`
if that helps

**Comment by bthirion** [comment]
> Mixed Gambles is really impractical and does not give strong result so I won't mind removing it but I think we should have something better instead if we want to showcase nilearn full scope.

Argh, yes.
We can provide a low-res IBC-RSVP example for within-subject decoding...
But it does not reduce the number of datasets :-(

**Comment by GaelVaroquaux** [comment]
> using requests needs approval from more people eg @GaelVaroquaux and @bthirion though, as it adds a new dependency

If it does make downloads more reliable then yes.

I thought that it was a small dependency, but it itself has dependencies.
However, it's probably already present in a huge fraction of our
userbase.

**Comment by jeromedockes** [comment]
ok I started adding requests and improving mocking. 
this is orthogonal to removing some datasets from examples and deprecating fetchers, which can be done in parallel

**Comment by Remi-Gau** [comment]
Would probably be a good thing to close this issue and open several other ones for the things that still need to be done on this front. Rereading the whole thread still leaves me wondering what is in the TODO and in the DONE camp.
